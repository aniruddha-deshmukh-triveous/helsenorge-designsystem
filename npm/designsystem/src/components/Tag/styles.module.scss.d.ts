export type Styles = {
  tag: string;
  tag__icon: string;
  'tag--banana': string;
  'tag--blueberry': string;
  'tag--cherry': string;
  'tag--emphasised': string;
  'tag--has-action': string;
  'tag--has-icon': string;
  'tag--kiwi': string;
  'tag--large': string;
  'tag--medium': string;
  'tag--neutral': string;
  'tag--normal': string;
  'tag--oncolor': string;
  'tag--plum': string;
  'tag--undo': string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
