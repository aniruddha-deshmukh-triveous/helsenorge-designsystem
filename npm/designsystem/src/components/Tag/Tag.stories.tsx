import React from 'react';

import { ComponentStory, ComponentMeta } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import Tag, { TagAction, TagSize, TagVariant } from './Tag';
import LawBook from '../Icons/LawBook';

export default {
  title: 'Components/Tag',
  component: Tag,
  argTypes: {
    children: {
      control: 'text',
      defaultValue: 'Tekst',
    },
    size: {
      control: 'select',
      options: TagSize,
      defaultValue: TagSize.medium,
    },
    color: {
      control: 'select',
      options: ['blueberry', 'neutral', 'cherry', 'banana', 'kiwi', 'plum'],
      defaultValue: 'blueberry',
    },
    variant: {
      control: 'select',
      options: TagVariant,
      defaultValue: TagVariant.normal,
    },
    action: {
      control: 'select',
      options: TagAction,
      defaultValue: TagAction.remove,
    },
  },
} as ComponentMeta<typeof Tag>;

export const Default: ComponentStory<typeof Tag> = (args: any) => (
  <div style={{ width: '40rem' }}>
    <Tag {...args}>{args.children}</Tag>
  </div>
);

export const WithIcon: ComponentStory<typeof Tag> = (args: any) => (
  <div style={{ width: '40rem' }}>
    <Tag {...args} svgIcon={LawBook}>
      {args.children}
    </Tag>
  </div>
);

export const Action: ComponentStory<typeof Tag> = (args: any) => (
  <div style={{ width: '40rem' }}>
    <Tag {...args} svgIcon={LawBook} onClick={action('Tag clicked')}>
      {args.children}
    </Tag>
  </div>
);
