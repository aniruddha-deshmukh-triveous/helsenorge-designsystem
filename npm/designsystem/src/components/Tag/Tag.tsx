import React from 'react';

import cn from 'classnames';

import { AnalyticsId } from '../../constants';
import { useHover } from '../../hooks/useHover';
import { palette, PaletteNames } from '../../theme/palette';
import Icon, { IconSize, SvgIcon } from '../Icons';
import Undo from '../Icons/Undo';
import X from '../Icons/X';

import styles from './styles.module.scss';

export enum TagSize {
  medium = 'medium',
  large = 'large',
}

export enum TagAction {
  remove = 'remove',
  undo = 'undo',
}

export enum TagVariant {
  normal = 'normal',
  oncolor = 'oncolor',
  emphasised = 'emphasised',
}

export type TagColors = Extract<PaletteNames, 'blueberry' | 'neutral' | 'cherry' | 'banana' | 'kiwi' | 'plum'>;

interface TagProps {
  /** Sets the text of the tag */
  children: string;
  /** Sets the size of the tag. Default: small */
  size?: keyof typeof TagSize;
  /** Sets the background of the tag. Not used if action is "undo". Default: blueberry */
  color?: TagColors;
  /** Adds an icon to the tag. Not shown if action is set. */
  svgIcon?: SvgIcon;
  /* Changes the appearance of the tag. Not used if action is "undo". Default: normal */
  variant?: keyof typeof TagVariant;
  /* Makes the tag a clickable button that performs an action. onClick must also be set. */
  action?: keyof typeof TagAction;
  /* Called when action is set and the tag is clicked on. action must also be set. */
  onClick?: () => void;
  /** Sets the data-testid attribute on the expander button. */
  testId?: string;
}

type ActionTagProps = Required<Pick<TagProps, 'children' | 'size' | 'color' | 'variant' | 'action' | 'onClick'>> & Pick<TagProps, 'testId'>;

const ActionTag: React.FC<ActionTagProps> = props => {
  const { children, size, color, variant, action, onClick, testId } = props;

  const { hoverRef, isHovered } = useHover<HTMLButtonElement>();

  const tagClasses = cn(
    styles.tag,
    styles[`tag--${size}`],
    styles[`tag--${action}`],
    styles['tag--has-action'],
    action === 'remove' && [styles[`tag--${color}`], styles[`tag--${variant}`]]
  );

  const getActionIcon = (): SvgIcon => {
    switch (action) {
      case 'remove':
        return X;
      case 'undo':
        return Undo;
    }
  };

  return (
    <button className={tagClasses} onClick={onClick} ref={hoverRef} type="button" data-testid={testId} data-analyticsid={AnalyticsId.Tag}>
      {children}
      <Icon
        svgIcon={getActionIcon()}
        size={IconSize.XXSmall}
        color={palette[`${action === 'undo' ? 'blueberry' : color}800`]}
        isHovered={isHovered}
      />
    </button>
  );
};

const Tag: React.FC<TagProps> = props => {
  const { children, size = TagSize.medium, color = 'blueberry', svgIcon, variant = 'normal', action, onClick, testId } = props;

  const tagClasses = cn(styles.tag, styles[`tag--${size}`], styles[`tag--${color}`], styles[`tag--${variant}`], {
    [styles['tag--has-icon']]: svgIcon,
  });

  if (action && onClick) {
    return (
      <ActionTag size={size} color={color} variant={variant} action={action} onClick={onClick} testId={testId}>
        {children}
      </ActionTag>
    );
  }

  return (
    <span className={tagClasses} data-testid={testId} data-analyticsid={AnalyticsId.Tag}>
      {svgIcon && <Icon svgIcon={svgIcon} size={IconSize.XXSmall} color={palette[`${color}800`]} className={styles.tag__icon} />}
      {children}
    </span>
  );
};

export default Tag;
