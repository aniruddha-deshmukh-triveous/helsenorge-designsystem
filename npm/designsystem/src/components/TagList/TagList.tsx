import React from 'react';

import styles from './styles.module.scss';

import { AnalyticsId } from '../../constants';

interface TagListPropsProps {
  /** Sets the data-testid attribute on the expander button. */
  testId?: string;
}

const TagList: React.FC<TagListPropsProps> = ({ children, testId }) => {
  return (
    <div className={styles['tag-list']} data-testid={testId} data-analyticsid={AnalyticsId.TagList}>
      {children}
    </div>
  );
};

export default TagList;
