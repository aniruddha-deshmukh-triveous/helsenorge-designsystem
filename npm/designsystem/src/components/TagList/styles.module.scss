@import '../../scss/_spacers.scss';
@import '../../scss/_breakpoints.scss';
@import '../../scss/_palette.scss';

.tag-list {
  display: inline-flex;
  align-items: center;
  flex-wrap: wrap;
  gap: getSpacer(2xs);
}
