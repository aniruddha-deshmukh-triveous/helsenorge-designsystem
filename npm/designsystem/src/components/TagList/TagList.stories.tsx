import React from 'react';

import { StoryObj, Meta } from '@storybook/react';

import TagList from './TagList';
import Docs from '../../docs';
import Tag from '../Tag';

const meta = {
  title: 'triveous-h-d-designsystem/Components/TagList',
  component: TagList,
  parameters: {
    docs: {
      page: (): React.JSX.Element => <Docs component={TagList} />,
      description: {
        component:
          'Som innbygger vil jeg kunne ta stilling til flere Tager med informasjon i en liste slik at jeg effektivt kan scanne gjennom nøkkelinformasjon for elementer i listen og gjennomføre mine oppgaver.',
      },
    },
  },
} satisfies Meta<typeof TagList>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  render: () => (
    <TagList>
      <Tag>{'Læring'}</Tag>
      <Tag>{'Psykisk helse'}</Tag>
      <Tag>{'Fysisk helse'}</Tag>
      <Tag>{'Livsstil'}</Tag>
    </TagList>
  ),
};
