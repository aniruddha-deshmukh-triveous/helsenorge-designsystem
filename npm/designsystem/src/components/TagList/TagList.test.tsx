import React from 'react';

import { render, screen, within } from '@testing-library/react';

import TagList from './TagList';
import Tag from '../Tag';

describe('Gitt at Tag skal vises vanlig', (): void => {
  describe('Når Tagen vises', (): void => {
    test('Så er teksten synlig', (): void => {
      render(
        <TagList>
          <Tag>Tekst</Tag>
          <Tag>Mer tekst</Tag>
        </TagList>
      );

      const list = screen.getByRole('list');
      expect(list).toBeVisible();

      const tagList = within(list).getAllByRole('listitem');
      expect(tagList).toHaveLength(2);
      expect(tagList[0]).toHaveTextContent('Tekst');
      expect(tagList[1]).toHaveTextContent('Mer tekst');
    });

    test('Så ser den ut slik den skal', (): void => {
      const { container } = render(
        <TagList>
          <Tag>Tekst</Tag>
          <Tag>Mer tekst</Tag>
        </TagList>
      );

      expect(container).toMatchSnapshot();
    });
  });
});
