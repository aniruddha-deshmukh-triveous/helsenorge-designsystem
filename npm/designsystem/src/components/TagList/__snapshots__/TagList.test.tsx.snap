// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Gitt at Tag skal vises vanlig > NÃ¥r Tagen vises > SÃ¥ ser den ut slik den skal 1`] = `
<div>
  <ul
    class="tag-list"
    data-analyticsid="tag-list"
  >
    <li
      class="tag-list__item"
    >
      <span
        class="tag tag--medium tag--blueberry tag--normal"
        data-analyticsid="tag"
      >
        Tekst
      </span>
    </li>
    <li
      class="tag-list__item"
    >
      <span
        class="tag tag--medium tag--blueberry tag--normal"
        data-analyticsid="tag"
      >
        Mer tekst
      </span>
    </li>
  </ul>
</div>
`;
